File: examples/go/message-board/MessageBoard_test.go
package contract

import (
	"testing"
	runar "github.com/icellan/runar/packages/runar-go"
)

func newBoard() *MessageBoard {
	return &MessageBoard{
		Message: runar.ByteString(""),
		Owner:   runar.Alice.PubKey,
	}
}

func TestMessageBoard_Post(t *testing.T) {
	b := newBoard()
	b.Post(runar.ByteString("hello"))
	if string(b.Message) != "hello" {
		t.Errorf("expected Message='hello', got %q", b.Message)
	}
}

func TestMessageBoard_PostMultiple(t *testing.T) {
	b := newBoard()
	b.Post(runar.ByteString("first"))
	b.Post(runar.ByteString("second"))
	if string(b.Message) != "second" {
		t.Errorf("expected Message='second', got %q", b.Message)
	}
}

func TestMessageBoard_Burn(t *testing.T) {
	b := newBoard()
	sig := runar.SignTestMessage(runar.Alice.PrivKey)
	b.Burn(sig)
}

func TestMessageBoard_BurnWrongKey_Fails(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure for wrong signature")
		}
	}()
	b := newBoard()
	sig := runar.SignTestMessage(runar.Bob.PrivKey)
	b.Burn(sig)
}

func TestMessageBoard_Compile(t *testing.T) {
	if err := runar.CompileCheck("MessageBoard.runar.go"); err != nil {
		t.Fatalf("Runar compile check failed: %v", err)
	}
}

func TestMessageBoard_OwnerUnchanged(t *testing.T) {
	b := newBoard()
	originalOwner := b.Owner
	b.Post(runar.ByteString("test"))
	if string(b.Owner) != string(originalOwner) {
		t.Error("expected Owner to remain unchanged after post")
	}
}

func TestMessageBoard_EmptyInitialMessage(t *testing.T) {
	b := &MessageBoard{Message: runar.ByteString(""), Owner: runar.Alice.PubKey}
	if string(b.Message) != "" {
		t.Errorf("expected empty message, got %q", b.Message)
	}
}

func TestMessageBoard_PostToEmpty(t *testing.T) {
	b := &MessageBoard{Message: runar.ByteString(""), Owner: runar.Alice.PubKey}
	b.Post(runar.ByteString("48656c6c6f"))
	if string(b.Message) != "48656c6c6f" {
		t.Errorf("expected '48656c6c6f', got %q", b.Message)
	}
}

File: examples/zig/message-board/MessageBoard_test.zig
const std = @import("std");
const root = @import("../examples_test.zig");
const runar = @import("runar");
const MessageBoard = @import("MessageBoard.runar.zig").MessageBoard;

const contract_source = @embedFile("MessageBoard.runar.zig");

test "compile-check MessageBoard.runar.zig" {
    const allocator = std.testing.allocator;
    try root.runar.compileCheckSource(allocator, contract_source, "MessageBoard.runar.zig");
}

test "MessageBoard init stores message and owner" {
    const owner = runar.ALICE.pubKey;
    const board = MessageBoard.init("48656c6c6f", owner);
    try std.testing.expectEqualSlices(u8, "48656c6c6f", board.message);
    try std.testing.expectEqualSlices(u8, owner, board.owner);
}

test "MessageBoard post updates message" {
    var board = MessageBoard.init("00", runar.ALICE.pubKey);
    board.post("48656c6c6f");
    try std.testing.expectEqualSlices(u8, "48656c6c6f", board.message);
}

test "MessageBoard burn succeeds with owner signature" {
    const board = MessageBoard.init("00", runar.ALICE.pubKey);
    board.burn(runar.signTestMessage(runar.ALICE));
}
